exist-core/src/main/java/org/exist/backup/restore/AppRestoreUtils.java

```diff
@@ -0,0 +1,180 @@
+package org.exist.backup.restore;
+
+import org.apache.commons.lang3.StringUtils;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.exist.Namespaces;
+import org.exist.backup.BackupDescriptor;
+import org.exist.repo.Deployment;
+import org.exist.repo.ExistRepository;
+import org.exist.storage.DBBroker;
+import org.exist.util.XMLReaderPool;
+import org.exist.xmldb.XmldbURI;
+import org.expath.pkg.repo.Package;
+import org.expath.pkg.repo.PackageException;
+import org.expath.pkg.repo.Packages;
+import org.expath.pkg.repo.deps.Semver;
+import org.xml.sax.Attributes;
+import org.xml.sax.InputSource;
+import org.xml.sax.SAXException;
+import org.xml.sax.XMLReader;
+import org.xml.sax.helpers.DefaultHandler;
+
+import java.io.IOException;
+import java.util.*;
+
+/**
+ * Utility to compare the applications contained in a backup with the already
+ * installed applications in the package repo.
+ *
+ * @author Wolfgang
+ */
+public class AppRestoreUtils {
+
+    private final static Logger LOG = LogManager.getLogger(AppRestoreUtils.class);
+
+    private final static String PKG_NAMESPACE = "http://expath.org/ns/pkg";
+
+    /**
+     * Inspects the apps contained in the backup against installed apps in the database
+     * and return a set of symbolic backup paths pointing to the collection of those
+     * apps for which newer versions are installed within the database. The returned
+     * paths may then be ignored during a restore.
+     *
+     * The method attempts to be fail safe to make sure even bad backups can be restored. Errors
+     * reading package descriptors are thus only logged and should not abort the process.
+     *
+     * @param broker the broker used for reading the backup and retrieving the expath repo
+     * @param descriptors a queue of backup descriptors to inspect
+     * @return a set of paths for which newer versions exist in the database. may be empty.
+     */
+    public static Set<String> checkApps(DBBroker broker, Deque<BackupDescriptor> descriptors) {
+        final List<AppDetail> apps = getAppsFromBackup(broker, descriptors);
+        final Set<String> paths = new HashSet<>();
+        final Optional<ExistRepository> repo = broker.getBrokerPool().getExpathRepo();
+        if (repo.isPresent()) {
+            for (final AppDetail app: apps) {
+                final Packages packages = repo.get().getParentRepo().getPackages(app.name);
+                if (packages != null) {
+                    final Package latest = packages.latest();
+                    try {
+                        final Semver version = Semver.parse(latest.getVersion());
```

Comment:
This will only work for apps which have valid SemVer 2.0 version numbers. @joewiz and I just fixed this elsewhere.
I am not certain, but I would that that could be a concern here too as there are some older packages with non-Semver 2.0 compatible naming?

Reply:
As far as I know the version attribute in the `expath-pkg.xml` always had to be a valid semver - thus the exception being thrown by Semver. Maybe you're confusing this with dependencies and `semver`, `semver-max` etc?

For invalid semvers no version check can be made, so corresponding packages can only be ignored by this code.

Reply:
actually expath failed to properly implement the intended checks. So while documentation said `semver` what it actually let through was not always proper semver. I fixed that a while ago.

Reply:
Also you can still do comparisons by coercion. Which is exactly what @joewiz and I implemented for the public-repo and packageservice.

@joewiz thoughts/comments please?

Reply:
for `semver` I'm aware there are no checks, but we're only dealing with `version` here. In any case it is unclear what the expected behaviour would be? We can assume that the eXist core apps which are required for proper operation have a valid semver. For user-supplied apps it is the users job to sort out the problem, which is easy if you have a working dashboard.

Reply:
I don't think we can make such assumptions - we should do the same coercion here...

Reply:
@wolfgangmm It looks like I might have already build coercion into the version you are using - https://github.com/expath/expath-pkg-java/blob/master/pkg-java/src/test/java/org/expath/pkg/repo/deps/SemverTest.java#L26

Maybe we are lucky? But we should add those two suggested tests around that